docsource/ops.rst

```diff
@@ -1,6 +1,228 @@
+      },
+      "dest": {
+        "index": "kbase.genome_1_r1",
+        "version_type": "external"
+      },
+      "script": {
+        "lang": "painless",
+        "inline": "if (ctx._source.foo == 'bar') {ctx._version++; ctx._source.remove('foo')}"
+      }
+    }
+
+6. Now, reindex the entire data from current index to new index. Alternately, use a query to reindex only a subset of the current index.
+
+.. code-block:: bash
+
+    POST _reindex
+    {
+      "source": {
+        "index": "kbase.genome_1"
+      },
+      "dest": {
+        "index": "kbase.genome_1_r1"
+      }
+    }
+
+        OR
+
+    POST _reindex
+    {
+      "source": {
+        "index": "kbase.genome_1",
+        "query": {
+          ...
+        }
+      },
+      "dest": {
+        "index": "kbase.genome_1_r1"
+      }
+    }
+
+7. Run a checksum on the new index to make sure the numbers line up with the numbers of the current index.
+
+.. code-block:: bash
+
+    GET /kbase.genome_1_r1/_stats/docs,store
+
+8. Run a query to specifically check the change that was applied.
+
+.. code-block:: bash
+
+    GET kbase.genome_1_r1/_search
+
+   OR
+
+    GET kbase.genome_1_r1/_search
+    {
+     "query": {
+       "match": {
+         "FIELD": "VALUE"
+       }
+     }
+    }
+
+   OR
+
+    https://www.elastic.co/guide/en/elasticsearch/reference/5.5/search-request-body.html
+
+9. If the new index looks good, update index alias and delete current index.
+
+.. note::
+
+    If you want the current index to linger for a day or two to serve a rollback option, reindex the current index into another new index called kbase.genome_1_backup and then delete the current index. This is one of two ways of renaming an index in ElasticSearch. The other way is to use the snapshot API.
+
+.. code-block:: bash
+
+    POST _aliases
+    {
+     "actions": [
+     {
+       "add": {
+         "index": "kbase.genome_1_r1",
+         "alias": "kbase.genome"
+         }
+       },
+       {
+         "remove": {
+         "index": "kbase.genome_1",
+         "alias": "kbase.genome"
+       }
+     }
+     ]
+    }
+
+    DELETE kbase.genome_1
+
+10. List all available indexes for the genome alias and all available genome indexes to ensure consistency across the alias map. Verify that all genome indexes that are present (except for backups) are referenced by the alias. Also verify that the alias does not contain an index reference for which no index exists.
+
+.. code-block:: bash
+
+    GET /_cat/aliases/kbase.genome
+
+    GET /_cat/indices/kbase.genome_*
+
+11. If the change involved in the reindexing operation also requires a corresponding search type spec change (located in resources/types/genome.yml for example), then this change must be applied.
+
+12. Change mapping version from "1" to "_r1" in the resources/types/genome.yml search type spec and add a comment (for future reference) that describes the change that took place in the r1 reindexing operation.
```

Comment:
Shouldn't we abstract away this detail from the search specs? From my understanding of the discussion with @scanon, the plan was to have aliases for each index so we could point `genome_1`, say, at the current index and then the specs don't need to be aware of the reindexing count, which makes things a lot simpler for the spec authors.

Reply:
How would this work? Right now the spec versions are keyed off their position in a list. Now a spec version will also have a reindexing version, right? So a spec maps to a specific index which consists of a type, a type version, and a reindexing version. Would the reindexing version be embedded in the spec? What if we reindex a different number of times in different environments?

Reply:
Great questions. Rather than keying off of the implicit position in a list, we can make the spec versions explicit. This also has the effect of making the specs a lot more user friendly for the spec authors. So the spec could look like - 

```
storage-type: WS
storage-object-type: KBaseGenomes.Genome
versions:
    -
        version: 1
        indexing-rules:
            -
                path: id
                keyword-type: keyword
     . . .
    -
        # a comment about what reindexing operation r1 was all about
        version: 1_r1
        indexing-rules:
            -
                path: id
                keyword-type: keyword
    . . .
```
Here, the reindexing version is represented as a minor version number of the spec type version. The reason for this is because I am looking at reindexing operations as hot-fixes to problems that will ultimately need to be addressed by the type specs and the indexer. For example, let's say we decide to reindex to change a field value. This reindex operation can be done fairly quickly in prod and is akin to a hot-fix as this change can potentially be made available to users almost immediately. But ultimately, the same field value change will need to be made in the type spec as well with a lookup transform in order to keep the index and spec in synch. The latter will take relatively more time (hopefully not a whole lot longer if we have a well defined workflow and no code changes/maintenance to take care of for the change) as it has to go through a slower workflow but will be required in order to continue making the trickle updates to the new reindexed index via the indexer. In the example above, the synching update to the type spec version 1 can be represented as version 1_r1. 

So the reindexing workflow I am proposing is as follows: 
1. stop trickle updates
2. reindex (not recommended, but potentially can go live right here)
3. update type spec
4. start trickle updates on old and new index (new index through version 1_r1)
5. if all looks good, update the alias
6. at some point in the future consolidate versions 1_r1, 1_r2, ..., 1_rn into a new version 2. 

This way every move we make on the index is documented, synched up with the formal specs and executed through a well defined workflow.

Regarding different envs, we need to maintain separete specs for the different envs. This is why I suggested the following dir structure. 
```
resources
   |_ ci
       |_ typemappings
       |_ types
   |_ appdev
       |_ typemappings
       |_ types
...
```

I think it will be best to maintain this structure in KBaseSearchEngine itself as the specs are pretty integral with the application and after thinking about it for a while, I tend to disagree that the specs can be viewed as data that needs to be separated from the application. I view them as configurations for the application for the various envs. In fact, I think we should change the resources dir name to config. Since we have a finite set of envs, we can maintain them easily in github alongside the application and solve a lot of process and communication related problems this way.